dns-cloudflare: Implement limited-scope API Tokens

Tugzrida opened the pull request:
A while ago Cloudflare added support for limited-scope API Tokens in place of using a global API key, but support for them in cloudflare/python-cloudflare took a while to get through.

In summary, this PR:
- ~~Adds the newer version of python-cloudflare with token support to the requirements~~
- Implements token functionality through the INI file parameter `dns_cloudflare_api_token` (in addition to the traditional `dns_cloudflare_email` and `dns_cloudflare_api_key`). This needed a more advanced parameter validator than the built in `required_variables` mechanism.
- Updates the docs to reflect the new option, needed token permissions, and version details of the `cloudflare` module

~~Testing for the plugin with tokens is semi-implemented, but at the moment requires switching between testing with a token and the traditional global key manually. I'm not experienced with `unittest` so I'll need some help with this.~~

~~Testing is mostly implemented, however coverage is failing on a few lines which I don't think can be tested.~~

Tox also seemed to have some issues running, even though I followed the developer guide.
(For future reference, this is because the file `/etc/letsencrypt/cli.ini` existed on my dev machine)

Fixes #7252 

## Pull Request Checklist

- [x] If the change being made is to a [distributed
  component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout),
  edit the `master` section of `CHANGELOG.md` to include a description of the
  change being made.
- [X] Add [mypy type
  annotations](https://certbot.eff.org/docs/contributing.html#mypy-type-annotations)
  for any functions that were added or modified.
- [X] Include your name in `AUTHORS.md` if you like.

alexzorin commented:
The tox failure is due to test coverage falling below the required threshold in `dns_cloudflare.py`. 

You can see which line numbers are missing test coverage [in the CI output](https://travis-ci.com/certbot/certbot/jobs/259343724#L785) (or by running `tox -e py27-cover` locally).

Once the unit tests are fixed up to test both kinds of authentication, it will probably resolve itself.

Tugzrida commented:
Tox locally isn't giving the same output as the CI sites - that's what's confusing me.

Figured out how to add some more unit tests and that's improved the coverage but theres a few lines that I'm not sure how to test(and were not covered in previous CI runs on other PRs)

alexzorin commented:
I'd forget about the missing 1% for now, can fix that after review. The important bits look OK :+1: .

Tugzrida commented:
I've resolved the merge conflict and improved coverage enough for the CI to pass.

Tox locally is still failing on [`test_renew_with_bad_int`](https://github.com/certbot/certbot/blob/master/certbot/tests/main_test.py#L1200) - throwing [`AssertionError: True is not false`](https://github.com/certbot/certbot/blob/master/certbot/tests/main_test.py#L1191) even on a completely clean clone of the repo. I must be missing something - otherwise I can open an issue if this should be working.

Is there anything I need to do to flag for review or just wait?

Tugzrida commented:
In reference to [this](https://github.com/certbot/certbot/issues/7252#issuecomment-512535999) comment from @bmw:

I have just tested my code with the minimum version of `cloudflare` currently required by certbot (which is 1.5.1, prior to the introduction of tokens in 2.3.1) using tokens and the Cloudflare API itself returns an error as the token is not presented correctly (as expected for versions of `cloudflare` prior to 2.3.1)

This is raised as a `CloudFlareAPIError` with code 6003 which is already caught [here](https://github.com/certbot/certbot/pull/7583/files#diff-8855bfb807e6e16e0939db86b401300aR177), so the message could be changed to also notify of a potentially old version, or `CloudFlare.__version__` could be used to check automatically.

Tugzrida commented:
Just fixed another merge conflict and corrected where I'd listed the needed permissions for the Cloudflare token.


The issue was that setting the `Zone:DNS:Edit` permissions on specific domains works fine for editing records, but to edit the records, you need the zone ID.

Asking the user to provide the ID, even though it's listed in the Cloudflare Dash, would IMO be a bit of a burden on the user (then also needing to store it for renewals, etc.), so therefore I think it'd be best to leave it the way it is, using the zone name to get the ID.

Obtaining the zone ID from the zone name requires much broader permissions for some reason, which makes the Tokens a lot less valuable as a scope-limiting measure, however they still have substantially limited scope over the Global Key, so I think it's still worth implementing them now and hoping Cloudflare refines their Tokens in the future. If anyone wants to push Cloudflare on that front - go ahead. I've had enough dealings with them in the past few months :joy:

This issue has also come up in implementing Cloudflare Tokens in lexicon: AnalogJ/lexicon#460

Depau commented:
Hi, I tested this and I can confirm it works.

It additionally requires the `Zone:Zone:Read` permission, though, that is not mentioned in the proposed documentation.

Tugzrida commented:
@Depau I'd already corrected the permissions in the docs, perhaps you're looking at an old commit. :)

FranklinYu commented:
@alexzorin Is there any news that you can share about this PR? Any thing the community can do to push this forward?

bmw commented:
> In reference to [this](https://github.com/certbot/certbot/issues/7252#issuecomment-512535999) comment from @bmw:
>
> I have just tested my code with the minimum version of `cloudflare` currently required by certbot (which is 1.5.1, prior to the introduction of tokens in 2.3.1) using tokens and the Cloudflare API itself returns an error as the token is not presented correctly (as expected for versions of `cloudflare` prior to 2.3.1)
>
> This is raised as a `CloudFlareAPIError` with code 6003 which is already caught [here](https://github.com/certbot/certbot/pull/7583/files#diff-8855bfb807e6e16e0939db86b401300aR177), so the message could be changed to also notify of a potentially old version, or `CloudFlare.__version__` could be used to check automatically.

Thanks for the PR and looking into this!

For the sake of simplicity, my recommendation would be to just update the hint to also say that if you're trying to use limited-scope API tokens, cloudflare>=x.y.z is required.

I would also mention this in the docs where I think we should still push people towards the limited tokens, but make sure it is clear that the global API tokens can be used if you're using something like an old distro package where cloudflare cannot be upgraded through your package manager.

Finally, I think we should keep the minimum version of cloudflare unchanged in `setup.py` and `tools/oldest_constraints.txt`.

Tugzrida commented:
Just fixed the conflict and reverted the versions in `setup.py` and `tools/oldest_constraints.txt`. Will work on the docs and update the error message shortly :)

Tugzrida commented:
Amended the documentation, let me know if it's too wordy.

Besides the version issue with `python-cloudflare` (which is now in the docs), the only remaining caveat is the Token needing quite permissive permissions.

As I mentioned a while ago, this could be improved at the cost of usability and simplicity by requiring the user to provide Zone IDs, but I feel that the correct way to remedy this is getting Cloudflare to create some token permission that allows the conversion of zone name to zone ID for individual zones rather than all zones on the account.

As this is more of a nuanced issue, me creating a support ticket with CF would probably be a bit difficult. If anyone has dev contacts, then that's probably the way to go.

bmw commented on certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py:
```diff
@@ -151,9 +174,14 @@ def _find_zone_id(self, domain):
                     hint = 'Did you enter the correct email address?'
+                elif code == 9109:
+                    hint = ('Does your API token have "Zone:Zone:Read" and "Zone:DNS:Edit" '
+                            'permissions for all zones?')
```
How do I set up my API token so I can hit this case to test it?

Tugzrida replied:
Should've been somewhere else, fixed now :)

Tugzrida commented:
Ah! I did have a `cli.ini` on my dev machine! Tests are running properly now.

I'll take a look at those permission problems - maybe CF has already fixed things! Working on the other bits too!

Tugzrida commented on certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py:
```diff
@@ -151,9 +174,14 @@ def _find_zone_id(self, domain):
-                    hint = 'Did you copy your entire API key?'
+                    hint = ('Did you copy your entire API token/key? To use Cloudflare tokens, '
+                            'you\'ll need python-cloudflare>=2.3.1. You have python-cloudflare {}'
+                            .format(CloudFlare.__version__))
```
I thought it was worth having as the version may not have caused that error, it may have been the length of the token/key, which is the reason that this error happens when using an older version of cloudflare. An older version would present a token as the global key to the server, but it's the wrong length.

Tugzrida commented on certbot/CHANGELOG.md:
```diff
@@ -33,7 +33,7 @@ More details about these changes can be found on our GitHub repo.
 ### Added
 
-*
+* Added support for Cloudflare's limited-scope API Tokens(certbot#7252)
```
Also moved it to the master section :)

Tugzrida commented:
That's the easy bits done. Will work on the permissions issues later today.

bmw commented:
Sounds good!

Sorry about the test failures. It's a known issue: https://github.com/certbot/certbot/issues/5201

Tugzrida commented:
sed 's/later today/tomorrow/'

@bmw it looks like that permission behaviour you're seeing is due to the guessing mechanism which recursively tries to find the base domain. I'm guessing that you tried to get a cert for `subdomain.example.com`, if `example.com` were the domain in your Cloudflare account. 

The logic would've first asked Cloudflare for the details of `subdomain.example.com` (intending to try `example.com` after), but would've failed as for some reason, despite the token having very permissive permissions, those permissions wouldn't seem to extend to asking about zones that don't exist 🤦‍♂

I don't even know what permission would need to be added to allow this to not fail and simply return an empty result as would happen when using the global key. I would've thought that the token already had `com.cloudflare.api.account.zone.list` as it is able to list existing domains.

This leaves basically four options:
- Ask the user to provide the zone ID for each domain they want a cert for, and store the IDs for use in renewals
- Ask the user to provide the base domain for each domain

(both of those are bad ideas imo)

- Ignore that error, which would miss an actual error(maybe tally up the errors and if every guess fails, then raise the error?)
- Get Cloudflare to improve Tokens, with a per-zone permission to convert zone names to IDs. This approach would also allow the token to be limited to individual zones, rather than needing all zones in the account.

I also think that case for code 9109 is a hangover from some testing I was trying. I'll wait to resolve it until this current issue is resolved. I'll probably also need to add some permission catches in other places.

bmw commented:
> @bmw it looks like that permission behaviour you're seeing is due to the guessing mechanism which recursively tries to find the base domain. I'm guessing that you tried to get a cert for subdomain.example.com, if example.com were the domain in your Cloudflare account.

You're right. I tried using the base domain and it worked.

> I'm leaning toward a combination of those last two options. 

I agree.

If the exception raised out of `zones.get` is one we currently recognize, I think we should continue to raise the exception, but otherwise, I think we should just log the exception and try the next guess for the base domain.

Kirill888 commented:
Just thought I'll share my "findings", I had success with the following token permissions:

Zone - DNS Edit, Zone:read (for specific zone)
Account - Access: Organizations, Identity Providers, and Groups:Read (for specific account)

I guess I would be happier with edit for specific zone with read for specific account, than edit for all zones but can't read account info, if I had mutiple zones that is...

Mind you I tested it with "slightly patched"  released code (dodgy edit to use token instead of email) rather than with code from this PR.

Tugzrida commented:
@Kirill888 That's interesting. I hadn't tried all of the options. It seems like CF has more granularity in the backend than they're presenting to the user when setting up tokens(they talk about `com.cloudflare.api.account.zone.list`, which isn't anywhere I can find)

The only problem I can see with that account permission is that the token would then have read access to details about Cloudflare's Access service, which is an identity management tool, and hence would probably be somewhat sensitive were it to be in use.

I'll have a more detailed look to see if there's a good compromise of options, but I think the best path is to get it working for now(which will still have stricter access than the global key) and then poke Cloudflare to expose the needed permissions in the UI, maybe even with a preset option like they have for Wordpress and other services.

Kirill888 commented:
@Tugzrida agree with you that having something out sooner, even without optimally tight permission boundary is still better than demanding the key to the whole account.

Ideally user would be able to, optionally, supply `zone id` in the `.ini` file, in which case all the zone guessing code won't need to run and the user can supply a token with least privilege, i.e. DNS:Edit for specific zone and nothing else.

It will then be up to the user to decide the balance between convenience and security. Obviously this will need to be documented in docs and in error messages, and that's the hard part.

- Most secure: `zone id` + `token: DNS:Edit (specific zone)`
- Ok secure: api `token` without `zone id` but with extra permissions to allow finding out `zone id`
   - DNS:Edit, Zone:Zone(read) all zones
   - OR: DNS:Edit, Zone:Zone(read) (specific zone) + Account: Read
- Least Secure: `email` + `global account key`

Tugzrida commented:
@Kirill888 that might be a good long-term solution, but I think in the short term it'll be easiest to simply get it working with the sub-optimal tokens and try to get CF to add an option for `com.cloudflare.api.account.zone.list` to the tokens UI, which would allow strictly permissioned tokens which work with the current logic.

Given that certbot is intended to make HTTPS simple and easy, I don't think that adding a bunch of different mechanisms is useful to that end, and would also make the codebase even more complicated. If an advanced user wanted such functionality, it'd probably be easier for them to adapt or write a hook script to do the challenge exactly as they wanted/needed for their application.

Tugzrida commented:
Should now ignore the silly domain matching token permission error. The error returned when this happens annoyingly has code `0` which is used by cloudflare for a bunch of other things, so I've caught on the permission name instead.

If no zone is found, then either the zone doesn't exist or the token actually does have the wrong permissions, so I've added to the error message for when no zone is found.

Also added some better error handling to the TXT record creation, as that is where a lack of DNS edit permissions will show up.

Tugzrida commented:
Made some minor improvements. Re-added the catch for `9109` in `find_zone_id` after some more detailed testing. It's triggered when an invalid token is used(string same length or longer than a token).

I don't know if I'm quite happy with using the error string in `find_zone_id` to ignore the incorrect token error. The other option would be only raising the error if `hint` has been set, however this would probably then not catch other errors from python-cloudflare like connectivity issues.

bmw commented on certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py:
```diff
@@ -170,22 +177,23 @@ def _find_zone_id(self, domain):
+                                     'that you have supplied valid Cloudflare API credentials.{2}'
+                                          .format(code, e, ' ({0})'.format(hint) if hint else ''))
 
             if zones:
```
I think we should set `zones` to an empty list before this `for` loop to match the case where `zones.get` returns an empty list when using an email address and API key. When testing this with an API token triggering the permissions error about `com.cloudflare.api.account.zone.list`, Certbot crashed with:

`UnboundLocalError: local variable 'zones' referenced before assignment`

bmw commented on certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py:
```diff
@@ -170,22 +177,23 @@ def _find_zone_id(self, domain):
-                raise errors.PluginError('Error determining zone_id: {0} {1}. Please confirm that '
-                                         'you have supplied valid Cloudflare API credentials.{2}'
-                                         .format(code, e, ' ({0})'.format(hint) if hint else ''))
+                if 'com.cloudflare.api.account.zone.list' not in str(e):
```
I have the same concern as you about whether or not we can rely on this string never changing. If it does, this plugin will break when requesting a cert for a subdomain.

Because of this, I suggest we:

1. Only raise an error here if it's an error code we currently recognize.
2. Log unraised errors through `logger.debug`.
3. If values like `code` are set, include information about the last raised exception in the error message raised at the bottom of this function.

bmw commented on certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py:
```diff
@@ -194,7 +202,10 @@ def _find_zone_id(self, domain):
+                                 'and is already associated with the supplied Cloudflare account. '
+                                 'If you\'re using Cloudflare Tokens, your Token needs '
+                                 '"Zone:DNS:Edit" and "Zone:Zone:Read" permissions for all zones '
+                                 'in your account.'
```
I think we should remove this new info about Cloudflare token permissions. As I said in another comment, I think we should include the last raised error message here and I think that including both that and this new info about token permissions is too wordy.

I'd personally prefer including the specific error now and if we're getting (m)any bug reports about people hitting this case and being unsure what permissions to use on their token, we can maybe update things here.

What do you think?

Tugzrida commented:
Sorry there's been a few bugs so far - I've just been testing my code in parts rather than all together. I've now got the pebble test server working and so should be able to get that feedback incorporated soon.

And just a few notes on the test server that don't seem to be covered in the developer guide:
- `certbot_test` will fail after restarting `run_acme_server` due to the account no longer existing. I worked around this by deleting the `.certbot_test_workspace` folder between restarts.
- The pebble server by default uses `127.0.0.1:8053` as its resolver. I haven't looked in great detail into how this is meant to work, but it was causing all my dns challenges to fail. I worked around this by changing the server referenced [here](https://github.com/certbot/certbot/blob/1702cb90fdc84e0f251df4d023b721a63f553011/certbot-ci/certbot_integration_tests/utils/acme_server.py#L134) to one of Cloudflare's authoritative servers, which worked great for my testing.

Tugzrida commented:
Just finished implementing those changes and tested fairly extensively against pebble. It worked for every case I could think of, so hopefully there aren't any hidden issues.

If there's an error that isn't recognised that prevents the obtaining of the zone ID, such as no network connection, It will throw something like this:
`certbot.errors.PluginError: Unable to determine zone_id for example.com using zone names: ['example.com', 'com']. Please confirm that the domain name has been entered correctly and is already associated with the supplied Cloudflare account. The error from Cloudflare was: 0 connection failed.`

About that error message in the final raise: I had put that there to clear up the possible cause of the problem. If the user's token does not include `Zone:Zone:Read`, this is what will now be thrown:
`certbot.errors.PluginError: Unable to determine zone_id for example.com using zone names: ['example.com', 'com']. Please confirm that the domain name has been entered correctly and is already associated with the supplied Cloudflare account. The error from Cloudflare was: 0 Actor 'com.cloudflare.api.token.<token>' requires permission 'com.cloudflare.api.account.zone.list' to list zones`

I think that's probably clear enough along with the docs, and as you say, it can be changed if it's causing problems in the future. Here you can also see the shared use of error code 0 by the cloudflare library which would prevent neat filtering on this.

I see that Travis has failed. I can correct the linting error(forgot to check it again before I pushed), but I'm not sure what to do about the mypy one. Should I initialise `zones` to `None` instead?

bmw commented:
Thanks for taking the time to write up your feedback about test issues. I created https://github.com/certbot/certbot/issues/7716 and https://github.com/certbot/certbot/issues/7717 to track those problems so hopefully others don't hit them in the future.

For the mypy issue, I'd apply this patch:
```
diff --git a/certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py b/certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py
index 592d347ac..c029d2dfe 100644
--- a/certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py
+++ b/certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py
@@ -4,6 +4,10 @@ import logging
 import CloudFlare
 import zope.interface
 
+from acme.magic_typing import Any
+from acme.magic_typing import Dict
+from acme.magic_typing import List
+
 from certbot import errors
 from certbot import interfaces
 from certbot.plugins import dns_common
@@ -169,7 +173,7 @@ class _CloudflareClient(object):
         """
 
         zone_name_guesses = dns_common.base_domain_name_guesses(domain)
-        zones = []
+        zones = []  # type: List[Dict[str, Any]]
         code = msg = None
 
         for zone_name in zone_name_guesses:
```
Other than that and the lint issue, this PR LGTM!

Tugzrida commented:
(looks like GPG had a little hiccup on that commit :/)

Sorry that took a while, I was waiting to get some time to do some final testing against pebble.

From a final look-over, the only thing I found was that the changelog entry was in an old section, so I've now fixed that.

The only other outstanding limitation is the non-ideal permissions needed on the Cloudflare Tokens.  As I said before, trying to go through public channels at CF to fix this may be difficult, so if anyone does have dev contacts there, then that would be useful :)

There were a few CF community threads that came up earlier here so I'll take a look at those in the morning and see how they're going.

bmw commented:
> From a final look-over, the only thing I found was that the changelog entry was in an old section, so I've now fixed that.

Nice catch. Thanks for fixing that.

This PR now LGTM. We're currently requiring 2 reviews on changes to our DNS plugins, so I'll try to wrangle someone else to give it a final review so we can get this merged. Thanks again for writing the PR and prodding Cloudflare to land the library changes upstream.

adferrand commented on certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py:
```diff
@@ -38,14 +42,35 @@ def more_info(self):  # pylint: disable=missing-docstring,no-self-use
+                                         'Global API Key. (see {})'
+                                         .format(credentials.confobj.filename, ACCOUNT_URL))
+        else:
+            raise errors.PluginError('{}: Either dns_cloudflare_api_token(recommended), or '
```
micro-nit: `dns_cloudflare_api_token (recommended)`

FranklinYu replied:
```suggestion
            raise errors.PluginError('{}: Either dns_cloudflare_api_token (recommended), or '
```

Tugzrida commented:
Just a note: Ubuntu 20.04 LTS starts freezes next Thursday, 27 Feb. [The current bundled certbot version looks to be 0.40.0](https://packages.ubuntu.com/focal/certbot), so it may be worth looking at pushing over the newer releases before the freeze.

bmw commented:
Thanks for pointing that out. It'll probably be tough to make the deadline as the package isn't even available in Debian unstable yet.

We have an [SRU exception](https://wiki.ubuntu.com/StableReleaseUpdates/Certbot) though and will likely be encouraging all of our users to switch to Certbot packaged as a snap in the near future. I suspect some users won't do this, but it'll be particularly easy for Ubuntu users and it gives them a way to always make sure they have the latest release installed.

FranklinYu commented:
1. If I understand correctly this feature is included in Certbot 1.2.0?
1. Is the Ubuntu package an/or Debian package maintained by anyone from the Certbot team?

Tugzrida commented:
Yes, tokens have made it into 1.2.0. Certbot in Ubuntu 20.04 is currently pulled from Debian, [which lists it as being maintained by the Debian Let's Encrypt team](https://packages.debian.org/sid/certbot)

Moving to snaps sounds great to allow more regular updates without messing around in PPA's and the like.

FranklinYu commented:
If I understand correctly this feature will be delivered in package [`python3-certbot-dns-cloudflare`](https://packages.debian.org/sid/python3-certbot-dns-cloudflare) instead of the package `certbot`. Package source is hosted on [GitLab](https://salsa.debian.org/letsencrypt-team/certbot/certbot-dns-cloudflare); I may poke them for release.

And yes, Snap could be a nice alternative that's officially supported. Would appreciate the effort even though I personally would use the Debian one.

bmw commented:
No one on the Certbot team is a Debian developer, but we maintain relationships with the volunteers who are that maintain these packages in their spare time.

@hlieberman, when you get a chance, I think it would be good to update the Cloudflare plugin in Debian to 1.2.0. It allows people to give Certbot fewer permissions on their Cloudflare account which has been a highly requested feature for us.

hlieberman commented:
Will take care of it tonight. Should be in sid by tomorrow and then
migration back to testing in the next five days or so.

On Mon, Feb 24, 2020 at 14:33 Brad Warren <notifications@github.com> wrote:

> No one on the Certbot team is a Debian developer, but we maintain
> relationships with the volunteers who are that maintain these packages in
> their spare time.
>
> @hlieberman <https://github.com/hlieberman>, when you get a chance, I
> think it would be good to update the Cloudflare plugin in Debian to 1.2.0.
> It allows people to give Certbot fewer permissions on their Cloudflare
> account which has been a highly requested feature for us.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/certbot/certbot/pull/7583?email_source=notifications&email_token=AAFNWXI4PK2XEEPQBAFJQXLREQOITA5CNFSM4JQMOQB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMZHJTI#issuecomment-590509261>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFNWXN4RT47OWM63WQWOLLREQOITANCNFSM4JQMOQBQ>
> .
>
-- 
Harlan Lieberman-Berg
~hlieberman

hlieberman commented:
```
Uploading python-certbot-dns-cloudflare using ftp to ftp-master (host:
ftp.upload.debian.org; directory: /pub/UploadQueue/)
running allowed-distribution: check whether a local profile permits uploads
to the target distribution
running protected-distribution: warn before uploading to distributions
where a special policy applies
running checksum: verify checksums before uploading
running suite-mismatch: check the target distribution for common errors
running gpg: check GnuPG signatures before the upload
 signfile dsc python-certbot-dns-cloudflare_1.2.0-1.dsc Harlan
Lieberman-Berg <hlieberman@debian.org>

 fixup_buildinfo python-certbot-dns-cloudflare_1.2.0-1.dsc
python-certbot-dns-cloudflare_1.2.0-1_amd64.buildinfo
 signfile buildinfo python-certbot-dns-cloudflare_1.2.0-1_amd64.buildinfo
Harlan Lieberman-Berg <hlieberman@debian.org>

 fixup_changes dsc python-certbot-dns-cloudflare_1.2.0-1.dsc
python-certbot-dns-cloudflare_1.2.0-1_source.changes
 fixup_changes buildinfo
python-certbot-dns-cloudflare_1.2.0-1_amd64.buildinfo
python-certbot-dns-cloudflare_1.2.0-1_source.changes
 signfile changes python-certbot-dns-cloudflare_1.2.0-1_source.changes
Harlan Lieberman-Berg <hlieberman@debian.org>

Successfully signed dsc, buildinfo, changes files
Uploading python-certbot-dns-cloudflare_1.2.0-1.dsc
Uploading python-certbot-dns-cloudflare_1.2.0.orig.tar.gz
Uploading python-certbot-dns-cloudflare_1.2.0.orig.tar.gz.asc
Uploading python-certbot-dns-cloudflare_1.2.0-1.debian.tar.xz
Uploading python-certbot-dns-cloudflare_1.2.0-1_amd64.buildinfo
Uploading python-certbot-dns-cloudflare_1.2.0-1_source.changes
```

On Mon, Feb 24, 2020 at 4:03 PM Harlan Lieberman-Berg <hlieberman@setec.io>
wrote:

> Will take care of it tonight. Should be in sid by tomorrow and then
> migration back to testing in the next five days or so.
>
> On Mon, Feb 24, 2020 at 14:33 Brad Warren <notifications@github.com>
> wrote:
>
>> No one on the Certbot team is a Debian developer, but we maintain
>> relationships with the volunteers who are that maintain these packages in
>> their spare time.
>>
>> @hlieberman <https://github.com/hlieberman>, when you get a chance, I
>> think it would be good to update the Cloudflare plugin in Debian to 1.2.0.
>> It allows people to give Certbot fewer permissions on their Cloudflare
>> account which has been a highly requested feature for us.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/certbot/certbot/pull/7583?email_source=notifications&email_token=AAFNWXI4PK2XEEPQBAFJQXLREQOITA5CNFSM4JQMOQB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMZHJTI#issuecomment-590509261>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAFNWXN4RT47OWM63WQWOLLREQOITANCNFSM4JQMOQBQ>
>> .
>>
> --
> Harlan Lieberman-Berg
> ~hlieberman
>


-- 
Harlan Lieberman-Berg
~hlieberman

bmw commented:
Thanks Harlan!

FranklinYu commented:
Thanks Harlan! I can already see it [in Sid](https://packages.debian.org/sid/python3-certbot-dns-cloudflare).

drzraf commented:
I think @Tugzrida workaround mentioned in https://github.com/certbot/certbot/pull/7583#issuecomment-575949323 stopped working during last days.

This does not work anymore:
![image](https://user-images.githubusercontent.com/1169926/75904117-620a5f00-5e21-11ea-98f0-d49611c1265a.png)

Tugzrida commented:
@drzraf I don't think I was ever able to get those permissions working. The permissions I've listed in the documentation (``Zone:Zone:Read`` and ``Zone:DNS:Edit`` for **all** zones) still work by my testing just now. Do those permissions work for you?

I'm keeping an eye on Cloudflare to see if and when these permissions can be tightened, but for now, the ones in the docs are the best compromise between security and usability.

drzraf commented:
Yes, **edit all zones** settings works for me, but I'm not satisfied with such a broad permission.
(https://community.cloudflare.com/t/trouble-with-token-for-clearing-cache-could-not-find-zone-id-for-zone/145462/10)

Tugzrida commented:
That is unfortunately the supported permissions at the time. The token will still be limited compared with the global key, but not as much as theoretically possible. In an ideal world, the token should only be able to add and remove records that match something like `^_acme-challenge\..*`, but Cloudflare doesn't support that fine of a permission control.

The best currently available would be DNS edit for specific zones, however that is not currently possible, as in order to get the Zone ID for a given zone name(the zone ID is necessary to edit DNS records), the token must have permissions to all zones. The workaround would be asking the user to manually copy the zone ID from their Cloudflare account to certbot. This would significantly impact usability for the average user, which is a problem seeing as the whole goal of Certbot/Let's Encrypt is to make HTTPS super easy.

The long term solution is asking Cloudflare to add a specific permission for Zone ID lookup. It looks like this may exist in the backend (`com.cloudflare.api.account.zone.list`), but it's not possible to select this in the tokens GUI.

In the meantime, I may make a hook script that would also work around the requirement for specific versions of the cloudflare python library. If I get this working, I'll publish a link here, but I'm not making any promises.

a-parser commented:
@Tugzrida instead of simply adding zone_id in ini config you keep talking on "best compromise between security and usability"

Sorry, but your work on limited-scope API Tokens are useless as you require DNS Edit access for ALL zones

Tugzrida commented:
For the average user of certbot, requiring extra work to manually add the zone ID's is a compromise to usability and therefore the goal of Certbot and Let's Encrypt.

API tokens are not useless as they still have significantly less permissions than the global API key. If you've previously been using certbot with a global key, using a token you will now be less vulnerable.

If your threat model and security policies cannot guarantee the security of your web server to the degree where an unideally-permissioned access token is a problem, then you are probably open to far greater attack vectors, and should probably consider a seperate server for certificate issuance, using a seperate domain via a CNAME for ACME verification.

Allowing for the manual entering and storage of zone IDs would require sacrifices to the simplicity of the user experience and significant additions to the codebase and therefore is not likely to happen. As I mentioned, you are more than welcome to write your own hook script to connect to the Cloudflare API in whatever way you want, but the current implementation in certbot is not likely to change until cloudflare adds better permissions to Tokens.

a-parser commented:
@Tugzrida it is frustrating that instead to provide more convenient and secure feature to end user you decided to spend much more time to proof your position, not the community

again, providing full access to editing DNS for all zones is absolutely dangerous and that's no what  "Limited-scope API Tokens" was designed for

Tugzrida commented:
I've explained that it would not be more convenient, and the security with the current token implementation is better than the global key anyway.

If you're not happy with the implementation, there are many other acme clients available, or you could write your own interface to the Cloudflare API.

I won't be commenting further.

mattjanssen commented:
> If your threat model and security policies cannot guarantee the security of your web server to the degree where an unideally-permissioned access token is a problem, then you are probably open to far greater attack vectors, and should probably consider a seperate server for certificate issuance, using a seperate domain via a CNAME for ACME verification.

For our case, our testing and staging servers are less secure than our production server; more people have access to the staging/testing secrets. Maybe we're doing it wrong? So if an API key from staging also gives access to our production domains, that would be un-ideal for us.

I suppose creating multiple CloudFlare accounts might be a solution. I'm not sure if we pay any per-account fee or not.

Tugzrida commented:
> So if an API key from staging also gives access to our production domains
That is a good point. I'm not sure about enterprise plans, but I'm pretty sure the Pro and Business plans are per-domain, so I'm fairly sure you can add the domains to additional accounts.

There may be changes we can make in the longer term if Cloudflare don't allow explicit access to the `com.cloudflare.api.account.zone.list` permission in the token GUI, but that really is the best path forward that I can see, so getting onto CF support(particularly if you have a paid plan with higher-priority support) and getting them to raise the issue is probably the best way on that front.

When I get time, I will be working on a seperate hook script for Cloudflare DNS auth, so that's an option too. During the process of writing it, I may stumble on a better way to do the whole thing(because honestly, relying on the whole `cloudflare` library is kinda silly for access to just a few endpoints)

ansman commented:
As a compromise perhaps the zone ID could be optional and, if not provided, requires the permission for all zones?

bmw commented:
> As a compromise perhaps the zone ID could be optional and, if not provided, requires the permission for all zones?

We may consider this in the future, but it's a little cumbersome because you could different domains with different zone IDs and right now I'd rather encourage Cloudflare to add a finer grained permission to get the zone ID.

ansman commented:
Ah, good point. I too think Cloudflare has a pretty bad permission model considering they are a security company

captainstatic commented:
Ugly but works with some restrictive permissions, allow all zone and exclude the ones you do not want. I agree, this should cleaned up on the Cloudflare side, but sometimes there are just "reasons" why it doesn't work that way.
https://community.cloudflare.com/t/how-to-get-zone-id-with-zone-specific-api-key/105300/2

Tugzrida commented:
This was mentioned in the new issue thread: https://github.com/certbot/certbot/issues/7893#issuecomment-615536801 :)

jsonUK commented:
FYI, installing the latest certbot via **snap** helped me get the latest version of the cdn-cloudflare plugin with the API tokens feature.

Tugzrida commented:
@jsonUK Yes, installing via snap is the recommended option for most platforms now as it will always be the latest version and stay up-to-date automatically.

For those that don't like snap, the OS packages still exist but may of course be a few versions behind due to the way your specific OS does releases.

Basic support for Cloudflare tokens starts in v1.2.0, with v1.6.0 being the most recent time as of writing this that any changes were made to the token behaviour.
